src/sources/kafka.rs

```diff
@@ -370,105 +411,559 @@ impl SourceConfig for KafkaSourceConfig {
+    // should completely drain its acknowledgement stream after receiving an end signal
+    let mut partition_consumers: JoinSet<(TopicPartition, PartitionConsumerStatus)> =
+        Default::default();
+
+    // Handles that will let us end any consumer task that exceeds a drain deadline
+    let mut abort_handles: HashMap<TopicPartition, tokio::task::AbortHandle> = HashMap::new();
+
+    let exit_eof = eof.is_some();
+
+    while let ConsumerState::Consuming(_) | ConsumerState::Draining(_) = consumer_state {
+        tokio::select! {
+            Some(Ok((finished_partition, status))) = partition_consumers.join_next(), if !partition_consumers.is_empty() => {
+                debug!("Partition consumer finished for {}:{}", &finished_partition.0, finished_partition.1);
+                // If this task ended on its own, the end_signal for it will still be in here.
+                end_signals.remove(&finished_partition);
+                abort_handles.remove(&finished_partition);
+
+                (drain_deadline, consumer_state) = match consumer_state {
+                    ConsumerState::Complete(_) => unreachable!("Partition consumer finished after completion."),
+                    ConsumerState::Draining(mut state) => {
+                        state.partition_drained(finished_partition);
+
+                        if state.is_drain_complete() {
+                            debug!("All expected partitions have drained.");
+                            state.finish_drain(drain_deadline)
+                        } else {
+                            state.keep_draining(drain_deadline)
+                        }
+                    },
+                    ConsumerState::Consuming(state) => {
+                        // If we are here, it is likely because the consumer
+                        // tasks are set up to exit upon reaching the end of the
+                        // partition.
+                        if !exit_eof {
+                            debug!("Partition consumer task finished, while not in draining mode.");
+                        }
+                        state.keep_consuming(drain_deadline)
+                    },
+                };
+
+                match status {
+                    // PartitionConsumerStatus differentiates between a task that exited after
+                    // being signaled to end, and one that reached the end of its partition and
+                    // was configured to exit. After the last such task ends, we signal the kafka
+                    // driver task to shut down the main consumer too. Note this is only used in tests.
+                    PartitionConsumerStatus::PartitionEOF if exit_eof && partition_consumers.is_empty() => {
+                        debug!("All partitions have exited or reached EOF.");
+                        let _ = eof.take().map(|e| e.send(()));
+                    },
+                    _ => {}
                 }
```

Comment:
While I like matches, this should probably just be an `if status == PartitionConsumerStatus::PartitionEOF && exit_eof && partition_consumers.is_empty()` (or an `if matches!(status, …)` if it can't be compared directly).